optuna/samplers/_tpe/multivariate_parzen_estimator.py

```diff
@@ -0,0 +1,375 @@
+
+            assert isinstance(distribution, _DISTRIBUTION_CLASSES)
+            if isinstance(distribution, distributions.UniformDistribution):
+                transformed[param_name] = samples
+            elif isinstance(distribution, distributions.LogUniformDistribution):
+                transformed[param_name] = np.exp(samples)
+            elif isinstance(distribution, distributions.DiscreteUniformDistribution):
+                q = self._q[param_name]
+                samples = np.round((samples - distribution.low) / q) * q + distribution.low
+                transformed[param_name] = np.clip(samples, distribution.low, distribution.high)
+            elif isinstance(distribution, distributions.IntUniformDistribution):
+                q = self._q[param_name]
+                samples = np.round(samples / q) * q
+                transformed[param_name] = np.clip(
+                    samples, distribution.low, distribution.high
+                ).astype(int)
+            elif isinstance(distribution, distributions.IntLogUniformDistribution):
+                samples = np.round(np.exp(samples))
+                transformed[param_name] = np.clip(
+                    samples, distribution.low, distribution.high
+                ).astype(int)
+            elif isinstance(distribution, distributions.CategoricalDistribution):
+                transformed[param_name] = samples
+
+        return transformed
+
+    def _precompute_sigmas0(self, multivariate_samples: Dict[str, np.ndarray]) -> np.ndarray:
+        # We use Scott's rule for bandwidth selection.
+        # This rule was used in the BOHB paper.
+        # TODO(kstoneriv3): The constant factor SIGMA0_MAGNITUDE=0.2 might not be optimal.
+        n_samples = next(iter(multivariate_samples.values())).size
+        n_samples = max(n_samples, 1)
+        n_params = len(multivariate_samples)
+        return SIGMA0_MAGNITUDE * n_samples ** (-1.0 / (n_params + 4)) * np.ones(n_samples)
+
+    def _calculate_categorical_params(
+        self, observations: np.ndarray, param_name: str
+    ) -> np.ndarray:
+
+        # TODO(kstoneriv3): This the bandwidth selection rule might not be optimal.
+        observations = observations.astype(int)
+        n_weights = self._n_weights
+        distribution = self._search_space[param_name]
+        assert isinstance(distribution, distributions.CategoricalDistribution)
+        choices = distribution.choices
+        consider_prior = self._parameters.consider_prior
+        prior_weights = self._parameters.prior_weight
+        if consider_prior:
+            shape = (n_weights + 1, len(choices))
+        else:
+            shape = (n_weights, len(choices))
+        weights = np.full(shape, fill_value=prior_weights / n_weights)
+        weights[np.arange(n_weights), observations] += 1
+        weights /= weights.sum(axis=1, keepdims=True)
+        return weights
+
+    def _calculate_parzen_est_params(
```

Comment:
Shouldn't this be `_calculate_numerical_params` instead?

Reply:
The '_calculate_numerical_params` is more reasonable to me than the `_calculate_parzen_est_params`.

Reply:
+1